filament/backend/src/webgpu/platform/WebGPUPlatform.cpp

```diff
@@ -67,72 +75,254 @@ namespace {
 
-    std::vector<wgpu::FeatureName> enabledFeatures;
-    enabledFeatures.reserve(requiredFeatures.size() + optionalFeatures.size());
-
-    std::set_intersection(supportedFeatures.features,
-            supportedFeatures.features + supportedFeatures.featureCount, requiredFeatures.begin(),
-            requiredFeatures.end(), std::back_inserter(enabledFeatures));
-
-    if (enabledFeatures.size() != requiredFeatures.size()) {
-        std::vector<wgpu::FeatureName> missingFeatures;
-        std::set_difference(requiredFeatures.begin(), requiredFeatures.end(),
-                supportedFeatures.features,
-                supportedFeatures.features + supportedFeatures.featureCount,
-                std::back_inserter(missingFeatures));
-
-        std::stringstream missingFeaturesStream{};
-        for (const auto& entry: missingFeatures) {
-            missingFeaturesStream << std::to_string(static_cast<uint32_t>(entry)) << " ";
+struct AdapterDetails final {
+    AdapterDetails()
+        : AdapterDetails(nullptr) {}
+    explicit AdapterDetails(wgpu::Adapter adapter)
+        : adapter(std::move(adapter)) {
+        info.nextInChain = &powerPreference;
+    }
+    AdapterDetails(wgpu::AdapterInfo&& info,
+            wgpu::DawnAdapterPropertiesPowerPreference& powerPreference, wgpu::Adapter&& adapter)
+        : info(std::move(info)),
+          powerPreference(powerPreference),
+          adapter(std::move(adapter)) {}
+    AdapterDetails& operator=(AdapterDetails&& other) noexcept {
+        adapter = std::exchange(other.adapter, nullptr);
+        info = std::exchange(other.info, {});
+        powerPreference = std::exchange(other.powerPreference, {});
+        return *this;
+    }
+
+    wgpu::AdapterInfo info{};
+    wgpu::DawnAdapterPropertiesPowerPreference powerPreference{};
+    wgpu::Adapter adapter = nullptr;
+
+    bool operator==(AdapterDetails const& other) const {
+        return info.vendorID == other.info.vendorID &&
+               info.deviceID == other.info.deviceID &&
+               info.backendType == other.info.backendType &&
+               info.adapterType == other.info.adapterType &&
+               powerPreference.powerPreference == other.powerPreference.powerPreference;
+    }
+};
+
+struct AdapterDetailsHash final {
+    size_t operator()(AdapterDetails const& details) const {
+        using utils::hash::combine;
+        size_t hash = 0;
+        combine(hash, details.info.vendorID);
+        combine(hash, details.info.deviceID);
+        combine(hash, static_cast<uint32_t>(details.info.backendType));
+        combine(hash, static_cast<uint32_t>(details.info.adapterType));
+        combine(hash, static_cast<uint32_t>(details.powerPreference.powerPreference));
+        return hash;
+    }
+};
+
+// trys to get a unique set of adapters with at least one adapter or panics if none are found
+[[nodiscard]] std::unordered_set<AdapterDetails, AdapterDetailsHash> requestUniqueAdapters(
```

Comment:
nit: unique isn't the right word here.  I would call this `getCompatibleAdapters`, that's more appropriate.  They should be unique always, but if we fail to get a compatible adapter, then we panic.  The same goes for the variable name `uniqueAdapters`. This is better called `compatibleAdapters`.